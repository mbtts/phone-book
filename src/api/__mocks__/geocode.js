const createResults = (lat, lang) => [
  {
    geometry: {
      location: {
        lat: () => lat,
        lng: () => lang
      }
    }
  }
];

const createMockGeocoder = (status, lat, lang) => {
  const results = createResults(lat, lang);
  return function() {
    return {
      geocode: (address, callback) => callback(results, status)
    };
  };
};

const initialiseMapsApi = jest.fn().mockImplementation(() => {
  return Promise.resolve({});
});

const geoCode = jest.fn().mockImplementation(() => {
  return Promise.resolve({
    lat: 53.6980027,
    lng: -2.476654
  });
});

export { geoCode, initialiseMapsApi, createMockGeocoder };
