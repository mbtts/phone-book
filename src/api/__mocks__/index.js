const mockData = [
  {
    id: 0,
    name: "Oleta Level",
    phone_number: "+442032960159",
    address: "10 London Wall, London EC2M 6SA, UK"
  },
  {
    id: 6,
    name: "Stanley Vanderhoof",
    phone_number: "+442032960000",
    address: "17 Anchor Ave, Darwen BB3 0AZ, UK"
  }
];

const api = jest.fn().mockImplementation(() => {
  return Promise.resolve(mockData);
});

export { api as default, mockData };
