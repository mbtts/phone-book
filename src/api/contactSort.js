const createLastName = contact => {
  const lastName = contact.name.match(/\S+/g).pop();
  return lastName;
};

const sortByLastName = modifier => {
  return (a, b) =>
    createLastName(a).localeCompare(createLastName(b)) * modifier;
};

const sortByLastNameAsc = sortByLastName(1);
const sortByLastNameDesc = sortByLastName(-1);

export { sortByLastNameAsc, sortByLastNameDesc };
