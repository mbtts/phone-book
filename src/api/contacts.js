const api = async () => {
  const result = await fetch(
    "https://www.mocky.io/v2/581335f71000004204abaf83"
  );
  const json = await result.json();
  const indexed = json.contacts.map((contact, i) => ({ id: i, ...contact }));
  return indexed.sort(sortByLastNameAsc);
};

const createLastName = contact => {
  const lastName = contact.name.match(/\S+/g).pop();
  return lastName;
};

const sortByLastName = modifier => {
  return (a, b) => {
    return createLastName(a).localeCompare(createLastName(b)) * modifier;
  };
};

const sortByLastNameAsc = sortByLastName(1);
const sortByLastNameDesc = sortByLastName(-1);

export { api as default, sortByLastNameAsc, sortByLastNameDesc };
