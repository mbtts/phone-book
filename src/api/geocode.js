const GLOBAL_INIT = "initMaps";

const cache = {};

const processResults = (address, results, resolve, reject) => {
  if (results[0]) {
    const location = results[0].geometry.location;
    const coords = { lat: location.lat(), lng: location.lng() };
    cache[address] = coords;
    resolve(coords);
  } else {
    reject("No results found");
  }
};

const geoCode = address => {
  return new Promise((resolve, reject) => {
    if (cache[address]) {
      return resolve(cache[address]);
    }

    const geocoder = new google.maps.Geocoder();
    geocoder.geocode({ address }, (results, status) => {
      status === "OK"
        ? processResults(address, results, resolve, reject)
        : reject(`Geocode error ${status}`);
    });
  });
};

const createInitialiser = resolve => () => {
  window[GLOBAL_INIT] = undefined;
  resolve(google);
};

const initialiseMapsApi = () => {
  return new Promise(resolve => {
    if (typeof google !== "undefined") {
      return resolve(google);
    }

    window[GLOBAL_INIT] = createInitialiser(resolve);
    const script = document.createElement("script");
    script.setAttribute(
      "src",
      `https://maps.googleapis.com/maps/api/js?key=${
        process.env.GOOGLE_MAPS_API
      }&callback=${GLOBAL_INIT}`
    );
    document.body.appendChild(script);
  });
};

export { geoCode, initialiseMapsApi };
