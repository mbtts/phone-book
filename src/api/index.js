const api = async () => {
  const result = await fetch("http://www.mocky.io/v2/581335f71000004204abaf83");
  const json = await result.json();
  return json.contacts;
};

export default api;
