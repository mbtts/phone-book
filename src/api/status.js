const status = {
  LOADING: "LOADING",
  LOADED: "LOADED",
  ERROR: "ERROR"
};

export default status;
