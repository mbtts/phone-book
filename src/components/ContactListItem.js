import { Link } from "react-router-dom";
import PropTypes from "prop-types";
import React from "react";

const ContactListItem = ({ contact }) => (
  <p>
    <Link to={`/user/${contact.id}`}>{contact.name}</Link>
  </p>
);

ContactListItem.propTypes = {
  contact: PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    phone_number: PropTypes.string.isRequired,
    address: PropTypes.string.isRequired
  }).isRequired
};

export default ContactListItem;
