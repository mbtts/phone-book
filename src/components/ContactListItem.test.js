import ContactListItem from "./ContactListItem";
import { MemoryRouter } from "react-router-dom";
import React from "react";
import { mockData } from "../api/__mocks__";

describe("Contact list component", () => {
  const contact = mockData[1];

  it("should render initial state", () => {
    const wrapper = shallow(<ContactListItem contact={contact} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should render the contact name and link", () => {
    const wrapper = shallow(
      <MemoryRouter>
        <ContactListItem className="listItem" contact={contact} />
      </MemoryRouter>
    );
    const link = wrapper
      .dive()
      .dive()
      .find("Link");

    expect(link.children().text()).toEqual(contact.name);
    expect(link.prop("to")).toEqual(`/user/${contact.id}`);
  });
});
