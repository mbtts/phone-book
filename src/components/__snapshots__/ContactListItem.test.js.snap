// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact list component should render initial state 1`] = `
<p
  className="listitem"
>
  <Link
    className="listitem__link"
    replace={false}
    to="/user/6"
  >
    Stanley Vanderhoof
  </Link>
</p>
`;
